pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,520 @@
+	var keyPEMBytes []byte
+	_, err := os.Stat(privateKeyPath)
+	if err == nil {
+		// load the private key contents from file
+		keyPEMBytes, err = ioutil.ReadFile(privateKeyPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to read key file %s: %v", privateKeyPath, err)
+		}
+	} else if err != nil && !os.IsNotExist(err) {
+		return nil, nil, fmt.Errorf("failed to stat key file %s: %v", privateKeyPath, err)
+	}
+	if len(keyPEMBytes) > 0 {
+		// try to parse private key from existing file
+		privateKey, err := keyutil.ParsePrivateKeyPEM(keyPEMBytes)
+		if err != nil {
+			klog.ErrorS(err, "Parse key from file error", "file", privateKeyPath)
+		} else {
+			return keyPEMBytes, privateKey, nil
+		}
+	}
+	// The key file does not exist or has been deleted. Generate a new key.
+	klog.Info("Generating new private key for IPSec")
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate new RSA private key: %v", err)
+	}
+	desiredKeyBytes, err := keyutil.MarshalPrivateKeyToPEM(key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal privatekey: %v", err)
+	}
+	if bytes.Equal(desiredKeyBytes, keyPEMBytes) {
+		return desiredKeyBytes, key, nil
+	}
+	klog.InfoS("Updating private key to file", "file", privateKeyPath)
+	if err = keyutil.WriteKey(privateKeyPath, desiredKeyBytes); err != nil {
+		return nil, nil, err
+	}
+	return desiredKeyBytes, key, nil
+}
+
+func loadCertificate(certPath string) ([]byte, []*x509.Certificate, error) {
+	var certPEMBytes []byte
+	_, err := os.Stat(certPath)
+	if err == nil {
+		// load the certificate from file
+		certPEMBytes, err = ioutil.ReadFile(certPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to read certificate file %s: %v", certPath, err)
+		}
+	} else if err != nil && !os.IsNotExist(err) {
+		return nil, nil, fmt.Errorf("failed to stat certificate file %s: %v", certPath, err)
+	}
+	if len(certPEMBytes) > 0 {
+		// try to parse the certificate from the existing file
+		certificates, err := certutil.ParseCertsPEM(certPEMBytes)
+		if err != nil {
+			klog.ErrorS(err, "Parse certificate from file error", "file", certPath)
+		} else {
+			return certPEMBytes, certificates, nil
+		}
+	}
+	return nil, nil, nil
+}
+
+func (c *Controller) configureOVS() error {
+	// calculate the hash of current configurations and save it to OVS DB to trigger updates.
+	hasher := sha256.New()
+	existing := c.certificateKeyPair.Load().(*certificateKeyPair)
+	hasher.Write(existing.rawCACert)
+	hasher.Write(existing.rawCert)
+	hasher.Write(existing.rawKey)
+	hash := hex.EncodeToString(hasher.Sum(nil))
+
+	ovsConfig := map[string]interface{}{
+		"certificate":      c.certPath,
+		"private_key":      c.privateKeyPath,
+		"ca_cert":          c.caPath,
+		"certificate_hash": hash,
+	}
+	oldConfig, err := c.ovsBridgeClient.GetOVSOtherConfig()
+	if err != nil {
+		return err
+	}
+	// delete the old config if needed
```

Comment:
Doesn't OVS support update?

Reply:
We have not implemented the update function for other_configs. But I think it is trivial to add one.

Reply:
It should be good to have, I assume it would be more seamless when the certificate updates

Reply:
Even before adding an atomic update func, we can put line 335 to 361 to OVS client as a single func called UpdateOVSOtherConfig()